Orange/widgets/utils/tests/concurrent_example.py

```diff
@@ -0,0 +1,147 @@
+                y_data[::2] = step % 2
+                y_data = np.random.rand(len(x_data))
+                # Needs a copy because projection should not be modified
+                # inplace. If it is modified inplace, the widget and the thread
+                #  hold a reference to the same object. When the thread is
+                # interrupted it is still modifying the object, but the widget
+                # receives it (the modified object) with a delay.
+                embedding = np.vstack((x_data, y_data)).T.copy()
+            step += 1
+            if step % (steps / 10) == 0:
+                state.set_progress_value(100 * step / steps)
+
+            if state.is_interruption_requested():
+                return res
+
+            res.embedding = embedding
+            state.set_partial_results(res)
+        return res
+
+
+class OWConcurrentWidget(OWDataProjectionWidget, ConcurrentWidgetMixin):
+    name = "Projection"
+    param = Setting(0)
+
+    def __init__(self):
+        OWDataProjectionWidget.__init__(self)
+        ConcurrentWidgetMixin.__init__(self)
+        self.embedding = None  # type: Optional[np.ndarray]
+
+    # GUI
+    def _add_controls(self):
+        box = gui.vBox(self.controlArea, True)
+        gui.comboBox(
+            box, self, "param", label="Parameter:",
+            items=["Param A", "Param B"], labelWidth=80,
+            callback=self.__param_combo_changed
+        )
+        self.run_button = gui.button(box, self, "Start", self._toggle_run)
+        super()._add_controls()
+
+    def __param_combo_changed(self):
+        super().start()
+
+    def _toggle_run(self):
+        if self.data is None:
+            return
+
+        # Pause task
+        if self.task is not None:
+            self.cancel()
+            self.run_button.setText("Resume")
+            self.commit()
+        # Resume task
+        else:
+            self.start()
+
+    # extend ConcurrentWidgetMixin
+    def _prepare_task(self, state: TaskState):
+        return partial(Runner.run, self.data,
+                       embedding=self.embedding, state=state)
+
+    def _set_partial_results(self, result: Results):
+        assert isinstance(result.embedding, np.ndarray)
+        assert len(result.embedding) == len(self.data)
+        first_result = self.embedding is None
+        self.embedding = result.embedding
+        if first_result:
+            self.setup_plot()
+        else:
+            self.graph.update_coordinates()
+            self.graph.update_density()
+
+    def _set_results(self, result: Results):
+        assert isinstance(result.embedding, np.ndarray)
+        assert len(result.embedding) == len(self.data)
+        super()._set_results(result)
+
+    def start(self):
+        self.run_button.setText("Stop")
+        super().start()
+
+    def on_done(self, future):
+        super().on_done(future)
+        self.run_button.setText("Start")
+        self.commit()
```

Comment:
I think setting the text and committing is something more appropriate for `_set_results`. `on_done` accepts a `Future`. Ideally, the main API wouldn't expose this kind of stuff to you. Maybe `on_done` should become private and have we'd have `_set_results` the only function that would handle when the task is complete.